internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +115,328 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+	}
+
+	if len(upstream.Spec.AllowAuthentication.Organizations.Allowed) > 0 {
+		return &metav1.Condition{
+			Type:    OrganizationsPolicyValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "Invalid",
+			Message: "spec.allowAuthentication.organizations.policy must be 'OnlyUsersFromAllowedOrganizations' when spec.allowAuthentication.organizations.allowed has organizations listed",
+		}, policy
 	}
 
-	// Sync loop errors:
-	// - Should not be configuration errors. Config errors a user must correct belong on the .Status
-	//   object. The controller simply must wait for a user to correct before running again.
-	// - Other errors, such as networking errors, etc. are the types of errors that should return here
-	//   and signal the controller to retry the sync loop. These may be corrected by machines.
-	return errorsutil.NewAggregate(errs)
+	return &metav1.Condition{
+		Type:    OrganizationsPolicyValid,
+		Status:  metav1.ConditionFalse,
+		Reason:  "Invalid",
+		Message: "spec.allowAuthentication.organizations.policy must be 'AllGitHubUsers' when spec.allowAuthentication.organizations.allowed is empty",
+	}, policy
 }
 
-func (c *gitHubWatcherController) validateUpstream(ctx controllerlib.Context, upstream *v1alpha1.GitHubIdentityProvider) (*upstreamgithub.ProviderConfig, error) {
-	result := upstreamgithub.ProviderConfig{
-		Name: upstream.Name,
+func (c *gitHubWatcherController) validateUpstreamAndUpdateConditions(ctx controllerlib.Context, upstream *v1alpha1.GitHubIdentityProvider) (
+	*upstreamgithub.ProviderConfig, // If validated, returns the config
+	error, // This error will only refer to programmatic errors such as inability to perform a Dial or dereference a pointer, not configuration errors
+) {
+	conditions := make([]*metav1.Condition, 0)
+	applicationErrors := make([]error, 0)
+
+	clientSecretCondition, clientID, clientSecret, clientSecretErr := c.validateClientSecret(upstream.Spec.Client.SecretName)
+	conditions = append(conditions, clientSecretCondition)
+	if clientSecretErr != nil {
+		applicationErrors = append(applicationErrors, clientSecretErr)
+	}
+
+	// Should there be some sort of catch-all condition to capture this?
+	// This does not actually prevent a GitHub IDP from being added to the cache.
+	groupNameAttribute, usernameAttribute, userAndGroupErr := validateUserAndGroupAttributes(upstream)
+	// Should not happen due to CRD defaulting
+	if userAndGroupErr != nil {
+		applicationErrors = append(applicationErrors, userAndGroupErr)
+	}
+
+	organizationPolicyCondition, policy := validateOrganizationsPolicy(upstream)
+	conditions = append(conditions, organizationPolicyCondition)
+
+	hostAndTLSConditions, host, httpClient, validateHostAndTLSErr := c.validateHostAndTLSConfiguration(ctx.Context, upstream.Spec.GitHubAPI)
+	conditions = append(conditions, hostAndTLSConditions...)
+	if validateHostAndTLSErr != nil {
+		applicationErrors = append(applicationErrors, validateHostAndTLSErr)
+	}
+
+	if len(conditions) != 5 {
```

Comment:
NOTE: untested?

Reply:
If this was triggered, its going to `log` the error and then `resync`.  But user input is required to fix this correct?  So we probably want a condition else a user is unlikely to know we hit this point?

Reply:
The thought was that if there are a wrong number of conditions it must be because some unexpected code path was traversed, which quite possibly is due to application error. In any case I'm not sure other controllers have this check, so maybe we don't need it here.